marlowe-actus/src/Language/Marlowe/ACTUS/Analysis.hs

```diff
@@ -92,3 +103,43 @@ genZeroRiskAssertions terms@ContractTerms{..} NpvAssertionAgainstZeroRiskBond{..
             in (constnt discountFactor) * (sign $ useval "payoff" t) + acc
         npv = foldl accumulateAndDiscount (constnt 0) (zip cfs [1..])
     in Assert (ValueLT (constnt expectedNpv) npv) continue
+
+getRiskFactors :: DataObserved -> EventType -> Day -> ContractTerms -> RiskFactors
+getRiskFactors dataObserved ev date ContractTerms{..} =
+  let riskFactors =
+        RiskFactors {
+          o_rf_CURS = 1.0
+        , o_rf_RRMO = 1.0
+        , o_rf_SCMO = 1.0
+        , pp_payoff = 0.0
+        }
+      observedKey =
+        case ev of
+          RR ->
+            ct_RRMO
+          SC ->
+            ct_SCMO
+          _  ->
+            ct_CURS
+      value =
+        case observedKey of
+          Just observedKey' ->
+            case M.lookup observedKey' dataObserved of
+              Just ValuesObserved{ values = values } ->
+                case L.find (\(ValueObserved { timestamp = timestamp }) -> timestamp == date) values of
+                  Just ValueObserved{ value = valueObserved } ->
+                    valueObserved
+                  Nothing ->
+                    1.0
+              Nothing ->
+                1.0
+          Nothing ->
+            1.0
```

Comment:
This could be written more concisely by taking advantage of the `Maybe` monad and the `do` notation. The whole `value` definition would become:
```haskell
value = fromMaybe 1.0 $ do observedKey' <- observedKey
                           ValuesObserved{ values = values } <- M.lookup observedKey' dataObserved
                           ValueObserved{ value = valueObserved } <- L.find (\ ValueObserved { timestamp = timestamp } -> timestamp == date) values
                           return valueObserved
```